File: medium/498_diagonal_traverse.rb
# 498. Diagonal Traverse
# Medium
#
# Given an m x n matrix mat, return an array of all the elements of the array
# in a zig-zag diagonal order.
#
# Example:
# [[1, 2, 3],
#  [4, 5, 6],
#  [7, 8, 9]]  -->  [1, 2, 4, 7, 5, 3, 6, 8, 9]

def find_diagonal_order(mat)
  rows = mat.size
  cols = mat[0].size
  res = []
  down = false

  (0...cols).each do |c|
    diag = []
    (0...[c + 1, rows].min).each do |r|
      diag << mat[r][c - r]
    end
    res << if down
             diag
           else
             diag.reverse
           end
    down = !down
  end

  (1...rows).each do |r|
    diag = []
    (0...[rows - r, cols].min).each do |c|
      diag << mat[r + c][cols - 1 - c]
    end
    res << if down
             diag
           else
             diag.reverse
           end
    down = !down
  end

  res.flatten
end

p find_diagonal_order([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
p find_diagonal_order([[1, 2], [3, 4]])
p find_diagonal_order([[1, 2, 3], [4, 5, 6]])
p find_diagonal_order([[1], [2], [3]])
p find_diagonal_order([[1], [2]])

File: python/498_diagonal_traverse.py
"""
498. Diagonal Traverse
Medium

Given an m x n matrix mat, return an array of all the elements of the array
in a zig-zag diagonal order.

Example:
[[1, 2, 3],
 [4, 5, 6],
 [7, 8, 9]]  -->  [1, 2, 4, 7, 5, 3, 6, 8, 9]
"""


class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        if len(mat) == 0:
            return []

        m = len(mat)
        n = len(mat[0])

        res = []
        for s in range(m + n - 1):
            tmp = []

            if s < n:
                i, j = 0, s
            else:
                i, j = s - n + 1, n - 1

            while i < m and j >= 0:
                tmp.append(mat[i][j])
                i += 1
                j -= 1

            if s % 2 == 0:
                tmp.reverse()

            res.extend(tmp)

        return res
